### ISSUE TITLE
Custom Dialog for Installer

### ISSUE THREAD

**Issue by SOliyhan** [oleg-shilo/wixsharp#1261]
Hi,
I want to use a custom dialog before the app installation and upgrade that will help me set some values like configuring proxy something else. How do I proceed with it using Custom UI. Thanks

**Comment by oleg-shilo** [comment]
You can create the WixSharp project from the "Custom UI" VS template. This will bring to the project source code for all dialogs in the UI sequence so you can customize it as you want.

**Comment by SOliyhan** [comment]
Hi,
I'm using Custom UI and created a dialog as per my requirements but how do I set public properties in property table. Since I'm taking a user input in the GUI so I guess properties regarding it will be required when I try to install using silent mode.

**Comment by oleg-shilo** [comment]
yes, you will need to decide on your default values for the properties for the silent mode as you no longer have UI to ask your user.

**Comment by oleg-shilo** [comment]
You are injecting the dialog correctly. it's not a problem. You can use "Custom Dialog" template as starting point .

The second question is not as easy to answer.

First of all MSI runs UI as not elevated. And I am not aware of any way to change it. This is done for a good reason. You are not modifying the system (installing files/registry keys) yet. You are only collecting the user input. You don't need elevation. When you proceed with the installation. MSI will ask the system for elevation.

Thus I encourage you to look at your deployment strategy and maybe rework it.

If you are still keen to do something "strong" from your process your can run an elevated external process from your custom dialog. This will do the trick.

**Comment by SOliyhan** [comment]
Thanks for the suggestion, you were right though and I am using AfterInstall now. So right now I am trying to get data for my `CustomDialog` from a file in the `InstallDir` during an upgrade with `project.BeforeInstall` to collect with `e.Data['OB']` and use this before it was removed for the new installation while upgrading in `project.AfterInstall` But I noticed in the logs that maybe while upgrading, uninstall(Before_install and After_Install) and re-install(Before_install and After_Install) are two different session with `Session.IsActive`. Moreover, I refered to [#75](https://github.com/oleg-shilo/wixsharp/issues/75) to get the `e.Data['OB']` from Before_install(uninstall) to AfterInstall(uninstall). How can I get this `e.Data['OB']` to use in AfterInstall(re-install) which is basically saving the session variable in the upgrade scenario.

**Comment by oleg-shilo** [comment]
The unfortunate answer is "you cannot". These are completely different sessions (uninstall->install) and MSI provides no persistence mechanism for that. :(

Thus your only options are either to save some temp file with data or to save data in the registry so it can be used across multiple sessions.

**Comment by SOliyhan** [comment]
Thanks, well there might be another approach to keep the specific file (not all) during upgrade and try to overwrite it in the install session. I'm not sure about it but I see msi installers preventing files from removing and rewriting it as per the need ?

**Comment by oleg-shilo** [comment]
You can use 
https://github.com/oleg-shilo/wixsharp/blob/e046cc13eaa34778c9b29113c06add22b0c8b625/Source/src/WixSharp.Samples/Wix%23%20Samples/Install%20Files/setup.cs#L35)

But it is rather a deployment antipattern.

**Comment by SOliyhan** [comment]
Hi Thanks for the help on the custom dialog, I have one final issue in silent mode for this custom dialog. On this dialog I uses a message box in the GUI whenever an error occurred. I only want to show this message box to users in GUI but not in silent mode so I tried `Session.UILevel()` but it is returning `UILevel=2 ` in GUI mode as well as in silent mode. I checked the docs and it says if I am using ManagedUI then UILevel will remain set to 2. That is why I am not able to distinguish between them, means how do I decide if I am not on silent mode then user see the dialog else don't show it.

**Comment by oleg-shilo** [comment]
Then indeed it is a limitation of the Managed UI architecture of MSI.
I don't have the answer. Maybe you can discover the parent process and check the command line params. 
Though it is extremely hacky.
